distreqx/bijectors/triangular_linear.py

```diff
@@ -0,0 +1,105 @@
+from ._linear import AbstractLinearBijector
+
+
+def _triangular_logdet(matrix: Array) -> Array:
+    """Computes the log absolute determinant of a triangular matrix."""
+    return jnp.sum(jnp.log(jnp.abs(jnp.diag(matrix))))
+
+
+class TriangularLinear(AbstractLinearBijector):
+    """A linear bijector whose weight matrix is triangular.
+
+    The bijector is defined as `f(x) = Ax` where `A` is a DxD triangular matrix.
+
+    The Jacobian determinant can be computed in O(D) as follows:
+
+    log|det J(x)| = log|det A| = sum(log|diag(A)|)
+
+    The inverse is computed in O(D^2) by solving the triangular system `Ax = y`.
+
+    The bijector is invertible if and only if all diagonal elements of `A` are
+    non-zero. It is the responsibility of the user to make sure that this is the
+    case; the class will make no attempt to verify that the bijector is
+    invertible.
+    """
+
+    _matrix: Array
+    _is_lower: bool
+
+    def __init__(self, matrix: Array, is_lower: bool = True):
+        """Initializes a `TriangularLinear` bijector.
+
+        Args:
+        matrix: a square matrix whose triangular part defines `A`. Can also be a
+            batch of matrices. Whether `A` is the lower or upper triangular part of
+            `matrix` is determined by `is_lower`.
+        is_lower: if True, `A` is set to the lower triangular part of `matrix`. If
+            False, `A` is set to the upper triangular part of `matrix`.
+        """
+        if matrix.ndim < 2:
+            raise ValueError(
+                f"`matrix` must have at least 2 dimensions, got {matrix.ndim}."
+            )
+        if matrix.shape[-2] != matrix.shape[-1]:
+            raise ValueError(
+                f"`matrix` must be square; instead, it has shape {matrix.shape[-2:]}."
+            )
+        super().__init__(event_dims=matrix.shape[-1])
+
+        self._matrix = jnp.tril(matrix) if is_lower else jnp.triu(matrix)
+        self._is_lower = is_lower
+
+    @property
+    def matrix(self) -> Array:
+        """The triangular matrix `A` of the transformation."""
+        return self._matrix
+
+    @property
+    def is_lower(self) -> bool:
+        """True if `A` is lower triangular, False if upper triangular."""
+        return self._is_lower
+
+    def forward(self, x: Array) -> Array:
+        """Computes y = f(x)."""
+        return self._matrix @ x
+
+    def forward_log_det_jacobian(self, x: Array) -> Array:
+        """Computes log|det J(f)(x)|."""
+        triangular_logdet = jnp.vectorize(_triangular_logdet, signature="(m,m)->()")
+        return triangular_logdet(self._matrix)
+
+    def forward_and_log_det(self, x: Array) -> Tuple[Array, Array]:
+        """Computes y = f(x) and log|det J(f)(x)|."""
+        return self.forward(x), self.forward_log_det_jacobian(x)
+
+    def inverse(self, y: Array) -> Array:
+        """Computes x = f^{-1}(y)."""
+        return jax.scipy.linalg.solve_triangular(self._matrix, y, lower=self.is_lower)
```

Comment:
This is a broader design question, but I'm open to input, should we depend on `lineax` (https://docs.kidger.site/lineax/)? Generally, I don't want more dependencies than necessary, but I also don't want redundant code that someone else is writing/maintaining. Not sure if we will use it enough to justify the dependency, but interested in others thoughts

Reply:
I'm not sure tbh, what would be the advantages in using lineax here? Compute time?

Reply:
Partially compute time (for different linear operators, they probably have very optimized solvers). Also for code reduction, just to outsource LoC. Given this is small enough for now we can just make it something to consider for later